tests/pools/meta/integration/test_rate_handling.py

```diff
@@ -0,0 +1,234 @@
+        base_swap,
+        swap,
+        wrapped_coins,
+        wrapped_decimals,
+        underlying_coins,
+        underlying_decimals,
+    ):
+        cls.alice = alice
+        cls.swap = swap
+        cls.base_swap = base_swap
+        cls.coins = wrapped_coins
+        cls.decimals = wrapped_decimals
+        cls.underlying_coins = underlying_coins
+        cls.underlying_decimals = underlying_decimals
+        cls.n_coins = len(wrapped_coins)
+
+    def setup(self):
+        # reset the virtual price between each test run
+        self.virtual_price_base = self.base_swap.get_virtual_price()
+        self.virtual_price = self.swap.get_virtual_price()
+
+        # approve base pool for swaps
+        base_coins = self.underlying_coins[self.n_coins - 1 :]
+        for idx in range(len(base_coins)):
+            base_coins[idx].approve(self.base_swap, 2 ** 256 - 1, {"from": self.alice})
+
+    def _min_max(self):
+        # get index values for the coins with the smallest and largest balances in the meta pool
+        balances = [self.swap.balances(i) / (10 ** self.decimals[i]) for i in range(self.n_coins)]
+        min_idx = balances.index(min(balances))
+        max_idx = balances.index(max(balances))
+        if min_idx == max_idx:
+            min_idx = abs(min_idx - 1)
+
+        return min_idx, max_idx
+
+    def _min_max_underlying(self, base=False):
+        # get index values for underlying coins with smallest and largest balances
+        balances = []
+        for i in range(len(self.underlying_coins)):
+            if i < self.n_coins - 1:
+                balances.append(self.swap.balances(i) / 10 ** self.underlying_decimals[i])
+            else:
+                base_i = i - (self.n_coins - 1)
+                balances.append(
+                    self.base_swap.balances(base_i) / (10 ** self.underlying_decimals[base_i])
+                )
+
+        if base:
+            for i in range(self.n_coins - 1):
+                balances.pop(0)
+
+        min_idx = balances.index(min(balances))
+        max_idx = balances.index(max(balances))
+        if min_idx == max_idx:
+            min_idx = abs(min_idx - 1)
+
+        return min_idx, max_idx
+
+    def rule_ramp_A(self, st_pct):
+        """
+        Increase the amplification coefficient.
+
+        This action happens at most once per test. If A has already
+        been ramped, a swap is performed instead.
+        """
+        if not hasattr(self.swap, "ramp_A") or self.swap.future_A_time():
+            return self.rule_exchange_underlying(st_pct)
+
+        new_A = int(self.swap.A() * (1 + st_pct))
+        self.swap.ramp_A(new_A, chain.time() + 86410, {"from": self.alice})
+
+    def rule_increase_rates(self, st_pct):
+        """
+        Increase the virtual price of the base pool.
+        """
+        if not hasattr(self.base_swap, "donate_admin_fees"):
+            # not all pools include `donate_admin_fees`
+            return self.rule_generate_fees()
```

Comment:
Pretty sure same case here as `ramp_A` - it will exist in all metapools.

Reply:
This is for the base pool, though. For instance, the `sbtc` pool does not have `donate_admin_fees()`. I'll update the comment accordingly

Reply:
Ahh yes, my mistake.